jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,441 @@
+
+    return learner_fn
+
+
+def learner_setup(env: Environment, config: Dict) -> Tuple[callable, RunnerState]:
+    """Initialise learner_fn, network, optimiser, environment and states."""
+    # Get available TPU cores.
+    cores_count = len(jax.devices())
+    # Get number of actions and agents.
+    num_actions = int(env.action_spec().num_values[0])
+    num_agents = env.action_spec().shape[0]
+    config["NUM_AGENTS"] = num_agents
+    # PRNG keys.
+    rng, rng_p = jax.random.split(jax.random.PRNGKey(config["SEED"]))
+
+    # Define network and optimiser.
+    network = ActorCritic(num_actions, config["ACTIVATION"])
+    optim = optax.adam(config["LR"])
+
+    # Initialise observation: Select only obs for a single agent.
+    init_x = env.observation_spec().generate_value()
+    init_x = jax.tree_util.tree_map(lambda x: x[0], init_x)
+    init_x = jax.tree_util.tree_map(lambda x: x[None, ...], init_x)
+
+    # initialise params and optimiser state.
+    params = network.init(rng_p, init_x)
+    opt_state = optim.init(params)
+
+    # Vmap network apply function over number of agents.
+    vmapped_network_apply_fn = jax.vmap(
+        network.apply, in_axes=(None, 1), out_axes=(1, 1)
+    )
+
+    # Get batched iterated update and replicate it to pmap it over cores.
+    learn = get_learner_fn(env, vmapped_network_apply_fn, optim.update, config)
+    learn = jax.pmap(learn, axis_name="device")
+
+    # Broadcast params and optimiser state to cores and batch.
+    broadcast = lambda x: jnp.broadcast_to(
+        x, (cores_count, config["UPDATE_BATCH_SIZE"]) + x.shape
+    )
+    params = jax.tree_map(broadcast, params)
+    opt_state = jax.tree_map(broadcast, opt_state)
+
+    # Initialise environment states and timesteps.
+    rng, *env_rngs = jax.random.split(
+        rng, cores_count * config["UPDATE_BATCH_SIZE"] * config["NUM_ENVS"] + 1
+    )
+    env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
+        jnp.stack(env_rngs),
+    )
+
+    # Split rngs for each core.
+    rng, *step_rngs = jax.random.split(
+        rng, cores_count * config["UPDATE_BATCH_SIZE"] + 1
+    )
+    # Add dimension to pmap over.
+    reshape_step_rngs = lambda x: x.reshape(
+        (cores_count, config["UPDATE_BATCH_SIZE"]) + x.shape[1:]
+    )
+    step_rngs = reshape_step_rngs(jnp.stack(step_rngs))
+    reshape_states = lambda x: x.reshape(
+        (cores_count, config["UPDATE_BATCH_SIZE"], config["NUM_ENVS"]) + x.shape[1:]
+    )
+    env_states = jax.tree_util.tree_map(reshape_states, env_states)
+    timesteps = jax.tree_util.tree_map(reshape_states, timesteps)
+
+    return learn, (params, opt_state, step_rngs, env_states, timesteps)
+
+
+def run_experiment(env: Environment, config: Dict) -> ExperimentOutput:
+    """Runs experiment."""
+    # Setup learner.
+    learn, initial_runner_state = learner_setup(env, config)
+
+    # Calculate total timesteps.
+    cores_count = len(jax.devices())
+    total_timesteps = (
+        cores_count
```

Comment:
Technically this returns how many CPU cores/GPUs/TPU cores so rather call it devices. As you could have 2 GPUs for example.

```suggestion
    n_devices = len(jax.devices())
    total_timesteps = (
        n_devices
```